LpcCrOSEC.p

```diff
@@ -271,33 +271,82 @@ NTSTATUS:comm_init_lpc() {
     return STATUS_SUCCESS;
 }
 
+/// Execute a command on the EC.
+///
+/// @param in [0] = version, [1] = command, [2] = outgoing length in bytes, [3] = incoming length in bytes, [4..38] = data
+/// @param in_size Must be 4 + ceil(outgoing length / 8)
+/// @param out [0] = number of received bytes or negative for error, [1..33] = data received
+/// @param out_size Must be 1 + ceil(incoming length / 8)
+/// @return An NTSTATUS
 DEFINE_IOCTL(ioctl_ec_command) {
-    if (in_size < 3 || out_size < 1)
+    if (in_size < 4)
         return STATUS_BUFFER_TOO_SMALL;
+    if (out_size < 1)
+        return STATUS_BUFFER_TOO_SMALL;
+
+    new NTSTATUS:status;
+    new version = in[0];
+    new command = in[1];
+    new ec_out_size = in[2];
+    new ec_in_size = in[3];
+    new in_data[EC_LPC_HOST_PACKET_SIZE];
+    new out_data[EC_LPC_HOST_PACKET_SIZE];
+    new result;
+
+    if (in_size < 4 + _ceil_div(ec_out_size, 8))
+        return STATUS_BUFFER_TOO_SMALL;
+    if (out_size < 1 + _ceil_div(ec_in_size, 8))
+        return STATUS_BUFFER_TOO_SMALL;
+
+    if (ec_out_size > EC_LPC_HOST_PACKET_SIZE - 1)
+        return STATUS_INVALID_PARAMETER;
+    if (ec_in_size > EC_LPC_HOST_PACKET_SIZE - 1)
+        return STATUS_INVALID_PARAMETER;
 
     if (ec_command_proto == 3) {
-        new command = (in[0] & 0xff) | (in[1] << 8);
-        new version = in[2];
-        new in_size_offset = 3;
-        new ec_in_size = in_size - in_size_offset;
-        new out_size_offset = 1;
-        new ec_out_size = out_size - out_size_offset;
-
-        return ec_command_lpc_3(command, version, in, ec_in_size, in_size_offset, out, ec_out_size);
+        unpack_bytes_le(in, in_data, ec_out_size, 4 * 8, 0);
+
+        status = ec_command_lpc_3(command, version, in_data, ec_out_size, out_data, ec_in_size, result);
+        if (status != STATUS_SUCCESS)
+            return status;
+
+        out[0] = result;
+        if (result >= 0)
+            pack_bytes_le(out_data, out, ec_in_size, 0, 8);
+        return STATUS_SUCCESS;
     }
 
     return STATUS_NOT_SUPPORTED;
 }
 
+/// Read memory from the EC memory map.
+/// @param in [0] = offset, [1] = bytes to read
+/// @param in_size Must be 2
+/// @param out [0..32] = data read from EC memory map
+/// @param out_size Must be at least 1 and equal ceil(bytes / 8)
+/// @return An NTSTATUS
 DEFINE_IOCTL(ioctl_ec_readmem) {
-    if (in_size < 1)
+    if (in_size < 2)
         return STATUS_BUFFER_TOO_SMALL;
+
     if (out_size < 1)
         return STATUS_BUFFER_TOO_SMALL;
 
+    new NTSTATUS:status;
     new offset = in[0] & 0xFFFF;
+    new bytes = in[1];
+    new out_data[EC_MEMMAP_SIZE];
 
-    return ec_readmem_lpc(offset, out_size, out);
+    if (out_size < _ceil_div(bytes, 8))
```

Comment:
exploitable overflow. if I set in[1] to 0x7FFFFFFFFFFFFFFF, this condition will pass but pack_bytes_le will attempt packing that amount of bytes, overwriting heap.

Reply:
Oh woops, I'm guessing `==` would fix that? Or should I change pack_bytes to `out_size * 8`?